qiskit_experiments/data_processing/nodes.py

```diff
@@ -596,3 +599,170 @@ class ProjectorType(Enum):
     ABS = ToAbs
     REAL = ToReal
     IMAG = ToImag
+
+
+class RestlessNode(DataAction, ABC):
+    """An abstract node for restless data processing nodes.
+
+    In restless measurements, the qubit is not reset after each measurement. Instead, the
+    outcome of the previous quantum non-demolition measurement is the initial state for the
+    current circuit. Restless measurements therefore require special data processing nodes
+    that are implemented as sub-classes of `RestlessNode`. Restless experiments provide a
+    fast alternative for several calibration and characterization tasks, for details
+    see https://arxiv.org/pdf/2202.06981.pdf.
+    """
+
+    def __init__(self, validate: bool = True):
+        """Initialize a restless node.
+
+        Args:
+            validate: If set to True the node will validate its input.
+        """
+        super().__init__(validate)
+        self._n_shots = None
+        self._n_circuits = None
+
+    def _format_data(self, data: Any) -> Any:
+        """Convert the data to an array.
+
+        This node will also set all the attributes needed to process the data such as
+        the number of shots and the number of circuits.
+
+        Args:
+            data: An array representing the memory.
+
+        Returns:
+            The data that has been processed.
+
+        Raises:
+            DataProcessorError: If the datum has the wrong shape.
+        """
+
+        self._n_shots = len(data[0])
+        self._n_circuits = len(data)
+
+        datum = np.array(data)
+
+        if self._validate:
+            if datum.shape != (self._n_circuits, self._n_shots):
+                raise DataProcessorError(
+                    f"The datum given to {self.__class__.__name__} does not convert "
+                    "of an array with dimension (number of circuit, number of shots)."
+                )
+
+        return data
+
+    def _reorder(self, unordered_data: np.array) -> np.array:
+        """Reorder the measured data according to the measurement sequence.
+
+        Here, is assumed that the inner loop of the measurement is done over the circuits
```

Comment:
This is hardcoded IBM-ish definition. For example, pulse simulator doesn't need to follow this (indeed it might be efficient to generate shot data first). Can we turn it into option like memory layout option of array (probably init arg of this node)
https://numpy.org/doc/stable/reference/generated/numpy.asarray.html

Reply:
I added an additional arg here: [ca56109](https://github.com/Qiskit/qiskit-experiments/pull/678/commits/ca561095cf47f6532f31ec54649f99b4b57dfed8). This now controls whether the single shot memory is transposed or not in the reorder method.